src/atlas/interpolation/element/Quad2D.cc

```diff
@@ -0,0 +1,229 @@
+    isect = T013.intersects(r, edgeEpsilon, epsilon);
+    if (isect) {
+        return isect;
+    }
+
+    Triag2D T231(v11, v01, v10);
+    isect = T231.intersects(r, edgeEpsilon, epsilon);
+    if (isect) {
+        isect.u = 1 - isect.u;
+        isect.v = 1 - isect.v;
+        return isect;
+    }
+
+    return isect.fail();
+}
+
+method::Intersect Quad2D::localRemap(const PointXY& p, double edgeEpsilon, double epsilon) const {
+    method::Intersect isect;
+
+    // work out if point is within the polygon
+    if (!inQuadrilateral({p.x(), p.y()}))
+        return isect.fail();
+
+    // calculate local coordinates via remapping technique
+    // if that fails fall back to finite-element scheme.
+    //
+    // ax**2 + bx + x = 0
+    auto solve_quadratic = [](const double a, const double b, const double c) {
+        double det = b * b - 4. * a * c;
+        Roots roots;
+        if (det >= 0.) {
+            double inv_two_a = 1. / (2. * a);
+            double sqrt_det = std::sqrt(det);
+            roots.a = (-b + sqrt_det) * inv_two_a;
+            roots.b = (-b - sqrt_det) * inv_two_a;
+        }
+        return roots;
+    };
+
+    // ax + b = 0
+    auto solve_linear = [](const double a, const double b) { return -b / a; };
+
+    auto validWeight = [&](const double w) { return (w > - epsilon) && (w < (1. + epsilon)); };
+
+    // solve for u and v where:
+    // w1 = ( 1 - u ) * ( 1 - v )
+    // w2 = u * ( 1 - v )
+    // w3 = u * v
+    // w4 = ( 1 - u ) * v
+
+    Vector2D ray(p.x(), p.y());
+    Vector2D vA = v00 - ray;
+    Vector2D vB = v10 - v00;
+    Vector2D vC = v01 - v00;
+    Vector2D vD = v00 - v10 - v01 + v11;
+
+    // solve for v
+    double a = cross2d(vC, vD);
+    double b = cross2d(vC, vB) + cross2d(vA, vD);
+    double c = cross2d(vA, vB);
+
+    if (std::abs(a) > epsilon) {
+        Roots roots = solve_quadratic(a, b, c);
+        if (validWeight(roots.a)) {
+            isect.v = roots.a;
+        }
+        else if (validWeight(roots.b)) {
+            isect.v = roots.b;
+        }
+        else {
+            isect.v = BAD_WEIGHT_VALUE;
+        }
+    }
+    else if (std::abs(b) > epsilon) {
+        isect.v = solve_linear(b, c);
+    }
+    else {
+        isect.v = BAD_WEIGHT_VALUE;
+    }
+
+    // solve for u
+    a = cross2d(vB, vD);
+    b = cross2d(vB, vC) + cross2d(vA, vD);
+    c = cross2d(vA, vC);
+
+    if (std::abs(a) > epsilon) {
+        Roots roots = solve_quadratic(a, b, c);
+        if (validWeight(roots.a)) {
+            isect.u = roots.a;
+        }
+        else if (validWeight(roots.b)) {
+            isect.u = roots.b;
+        }
+        else {
+            isect.u = BAD_WEIGHT_VALUE;
+        }
+    }
+    else if (abs(b) > epsilon) {
+        isect.u = solve_linear(b, c);
+    }
+    else {
+        isect.u = BAD_WEIGHT_VALUE;
+    }
+
+    if ((isect.u == BAD_WEIGHT_VALUE) || (isect.v == BAD_WEIGHT_VALUE)) {
+        return isect.fail();
+    }
+    else {
```

Comment:
Don't use else after return (although you might want to re-use the u and v logic, so the point is moot)